### PR TITLE
Add support for adding/removing inspections from a quick fix dynamically

### PR DESCRIPTION
**Pull request by Hosch250**
More C# 7 stuff
Clean up the inspection tests

### CODE REVIEW

File: Rubberduck.Inspections/Abstract/QuickFixBase.cs [hunk 1]
```diff
@@ -6,19 +6,36 @@
...
         }
 
-        public void RegisterInspections(params IInspection[] inspections)
+        public void RegisterInspections(params Type[] inspections)
         {
-            _supportedInspections = inspections.Where(w => w != null).Select(s => s.Type).ToHashSet();
+            if (!inspections.All(s => s.GetInterfaces().Any(a => a == typeof(IInspection))))
+            {
+                throw new ArgumentException($"Parameters must implement {nameof(IInspection)}", nameof(inspections));
+            }
```
Comment (Vogel612) [inline Rubberduck.Inspections/Abstract/QuickFixBase.cs]:
```
_supportedInspections = inspections.Where(s => s.GetInterfaces().Any(i => i == typeof(IInspection)))
inspections.Where(s => !s.GetInterfaces().Any(i => i == typeof(IInspection)).ForEach(/* log an error-level entry */)
```

**Comment by retailcoder** [comment]
@Hosch250 uh-oh, @Vogel612's work doesn't quite merge well with this one...

File: Rubberduck.Inspections/Abstract/QuickFixBase.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+            {
+#if DEBUG
+                throw new ArgumentException($"Parameters must implement {nameof(IInspection)}", nameof(inspections));
+#else
+                inspections.Where(s => s.GetInterfaces().All(i => i != typeof(IInspection))).ToList()
+                    .ForEach(i => _logger.Error($"Type {i.Name} does not implement {nameof(IInspection)}"));
+#endif
+            }
+
+            _supportedInspections = inspections.ToHashSet();
```
Comment (bclothier) [inline Rubberduck.Inspections/Abstract/QuickFixBase.cs]: Now that you're logging as per @Vogel612's suggestion, I wonder if the `RegisterInspection` should at least communicate whether it encountered an error so the calling code won't have incorrect assumptions about whether it was successful in registering the inspections?

Comment (Vogel612) [reply Rubberduck.Inspections/Abstract/QuickFixBase.cs]: @bclothier the calling code doesn't make any assumptions about the success of that operation

Comment (bclothier) [reply Rubberduck.Inspections/Abstract/QuickFixBase.cs]: I might have lost the plot but if the point is enabling users to load their custom inspections dynamically, among with those that ships with RD, doesn't it imply a need to message to the users there was a problem in registering the custom inspections? The message can refer to the log but the user has to at least know, no?

Comment (Vogel612) [reply Rubberduck.Inspections/Abstract/QuickFixBase.cs]: @bclothier the user will know as soon as they attempt to execute their custom inspections and quickfixes. Anything else would be premature and only relevant to a minuscle fraction of the userbase.

Who's going to develop their own inspections and quickfixes for rubberduck? And if there's only few of them, why should we "burden" the bulk of our users with diagnostic information for this "rare" case?

Comment (Hosch250) [reply Rubberduck.Inspections/Abstract/QuickFixBase.cs]: If they test it in Debug mode, as they should, they get an exception.

**Pull request merged**